In [1]:
# import necessary modules
import pandas as pd
import numpy as np 
import seaborn as sns
import matplotlib.pyplot as plt


from IPython.display import display

### Load the data & preview the data

In [2]:
data = pd.read_csv('../data/raw/Copy of Ask A Manager Salary Survey 2021 (Responses) - Form Responses 1.csv')

In [3]:
print(data.info())
data.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28061 entries, 0 to 28060
Data columns (total 18 columns):
 #   Column                                                                                                                                                                                                                                Non-Null Count  Dtype  
---  ------                                                                                                                                                                                                                                --------------  -----  
 0   Timestamp                                                                                                                                                                                                                             28061 non-null  object 
 1   How old are you?                                                                                             

Timestamp How old are you?  What industry do you work in?  \
0  4/27/2021 11:02:10            25-34   Education (Higher Education)   
1  4/27/2021 11:02:22            25-34              Computing or Tech   
2  4/27/2021 11:02:38            25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41            25-34                     Nonprofits   
4  4/27/2021 11:02:42            25-34  Accounting, Banking & Finance   

                                  Job title  \
0        Research and Instruction Librarian   
1  Change & Internal Communications Manager   
2                      Marketing Specialist   
3                           Program Manager   
4                        Accounting Manager   

  If your job title needs additional context, please clarify here:  \
0                                                NaN                 
1                                                NaN                 
2                                                NaN                 
3                                                NaN                 
4                                                NaN                 

  What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)  \
0                                             55,000                                                                                                                                                                                     
1                                             54,600                                                                                                                                                                                     
2                                             34,000                                                                                                                                                                                     
3                                             62,000                                                                                                                                                                                     
4                                             60,000                                                                                                                                                                                     

   How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.  \
0                                                0.0                                                                                                                                                
1                                             4000.0                                                                                                                                                
2                                                NaN                                                                                                                                                
3                                             3000.0                                                                                                                                                
4                                             7000.0                                                                                                                                                

  Please indicate the currency  \
0                          USD   
1                          GBP   
2                          USD   
3                          USD   
4                          USD   

  If "Other," please indicate the currency here:   \
0                      

### Cleaning data:
- standardize column names
- Remove unnecessary columns
- convert columns to the appropriate data type
- handle missing values

In [4]:
# Make a copy of the original data
df = data.copy()

In [5]:
df.columns

Index(['Timestamp', 'How old are you?', 'What industry do you work in?',
       'Job title',
       'If your job title needs additional context, please clarify here:',
       'What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)',
       'How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.',
       'Please indicate the currency',
       'If "Other," please indicate the currency here: ',
       'If your income needs additional context, please provide it here:',
       'What country do you work in?',
       'If you're in the U.S., what state do you work in?',
       'What city do you work in?',
       'How many years of professional work experience do you have overall?',
       

**Standardize column names**

In [6]:
column_names = {
    'Timestamp': 'timestamp',
    'How old are you?': 'age_range',
    'What industry do you work in?': 'industry',
    'Job title': 'job_title',
    'If your job title needs additional context, please clarify here:': 'job_title_context',
    "What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)":'salary(annualized)',
    'How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.': 'additional_compensation',
    'Please indicate the currency':'currency',
    'If "Other," please indicate the currency here: ': 'other_currency',
    'If your income needs additional context, please provide it here:': 'income_context',
    'What country do you work in?': 'country',
    "If you're in the U.S., what state do you work in?":'us_state',
    'What city do you work in?': 'us_city',
    'How many years of professional work experience do you have overall?': 'years_of_experience',
    'How many years of professional work experience do you have in your field?': 'years_of_experience_in_field',
    'What is your highest level of education completed?': 'level_of_education',
    'What is your gender?': 'gender',
    'What is your race? (Choose all that apply.)': 'race'
}
df.columns = df.columns.map(column_names)

In [7]:
df.head()

timestamp age_range                       industry  \
0  4/27/2021 11:02:10     25-34   Education (Higher Education)   
1  4/27/2021 11:02:22     25-34              Computing or Tech   
2  4/27/2021 11:02:38     25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41     25-34                     Nonprofits   
4  4/27/2021 11:02:42     25-34  Accounting, Banking & Finance   

                                  job_title job_title_context  \
0        Research and Instruction Librarian               NaN   
1  Change & Internal Communications Manager               NaN   
2                      Marketing Specialist               NaN   
3                           Program Manager               NaN   
4                        Accounting Manager               NaN   

  salary(annualized)  additional_compensation currency other_currency  \
0             55,000                      0.0      USD            NaN   
1             54,600                   4000.0      GBP            NaN   
2             34,000                      NaN      USD            NaN   
3             62,000                   3000.0      USD            NaN   
4             60,000                   7000.0      USD            NaN   

  income_context         country        us_state      us_city  \
0            NaN   United States   Massachusetts       Boston   
1            NaN  United Kingdom             NaN    Cambridge   
2            NaN              US       Tennessee  Chattanooga   
3            NaN             USA       Wisconsin    Milwaukee   
4            NaN              US  South Carolina   Greenville   

  years_of_experience years_of_experience_in_field level_of_education  \
0           5-7 years                    5-7 years    Master's degree   
1        8 - 10 years                    5-7 years     College degree   
2         2 - 4 years                  2 - 4 years     College degree   
3        8 - 10 years                    5-7 years     College degree   
4        8 - 10 years                    5-7 years     College degree   

       gender   race  
0       Woman  White  
1  Non-binary  White  
2       Woman  White  
3       Woman  White  
4       Woman  White

**Dropping Unnecessary Columns Or Columns with High Missing Values**

In [8]:
# drop columns with high missing values
(df.isnull().mean() *100).sort_values(ascending=False)

other_currency                  99.269449
income_context                  89.159331
job_title_context               74.120666
additional_compensation         25.996935
us_state                        17.896725
level_of_education               0.791134
race                             0.630769
gender                           0.609387
us_city                          0.292221
industry                         0.263711
job_title                        0.003564
timestamp                        0.000000
age_range                        0.000000
salary(annualized)               0.000000
currency                         0.000000
country                          0.000000
years_of_experience              0.000000
years_of_experience_in_field     0.000000
dtype: float64

'income_context', 'other_currency', 'job_title_context' all have more than 70% missing values, so we can drop them. But before doing so, let's check if there are any meaningful information in these columns.

In [9]:
print('Number of Non-null values in the Other_currency Column: ',len(df[~df['other_currency'].isnull()]))
print('Number of unique values in the Other_currency Column: ',df[~df['other_currency'].isnull()]['other_currency'].nunique())
df[~df['other_currency'].isnull()].iloc[:,:10].sample(20)

Number of Non-null values in the Other_currency Column:  205
Number of unique values in the Other_currency Column:  121


timestamp age_range  \
26727  8/26/2021 15:51:01     35-44   
26531   7/6/2021 18:49:41     25-34   
766    4/27/2021 11:14:39     25-34   
752    4/27/2021 11:14:29     45-54   
11760   4/28/2021 4:46:46     35-44   
4971   4/27/2021 12:53:00     35-44   
18930   4/29/2021 5:20:01     25-34   
7739   4/27/2021 15:09:39     45-54   
28054  7/12/2024 18:24:48     25-34   
9981   4/27/2021 19:33:22     55-64   
20097  4/29/2021 12:44:46     25-34   
26343   5/26/2021 6:48:44     35-44   
27630   1/24/2022 0:18:39     25-34   
27834  6/15/2022 10:10:04     35-44   
14331  4/28/2021 14:25:40     45-54   
19323   4/29/2021 8:59:21     35-44   
6969   4/27/2021 14:18:47     25-34   
4264   4/27/2021 12:29:22     25-34   
10948  4/27/2021 22:50:41     35-44   
8650   4/27/2021 16:25:55     45-54   

                                        industry  \
26727               Education (Higher Education)   
26531                                 Nonprofits   
766                            Recruitment or HR   
752    Public Library (technically City Govt.?)    
11760              Education (Primary/Secondary)   
4971                Engineering or Manufacturing   
18930                          Computing or Tech   
7739                                 Health care   
28054              Accounting, Banking & Finance   
9981                     Agriculture or Forestry   
20097                          Computing or Tech   
26343               Engineering or Manufacturing   
27630                          Computing or Tech   
27834                          Computing or Tech   
14331                                  Military    
19323                     Business or Consulting   
6969                           Computing or Tech   
4264                                         Law   
10948              Accounting, Banking & Finance   
8650                           Computing or Tech   

                                               job_title  \
26727                                       Receptionist   
26531                   Districtwide Program Coordinator   
766                   Lead Payroll & Benefits Specialist   
752                             Administrative Librarian   
11760                                        ESL Teacher   
4971                           Compliance Senior Analyst   
18930                                        GIS Analyst   
7739                          Board certified physician    
28054                                  Financial Analyst   
9981                                Executive Assistant    
20097                                         Consultant   
26343                                 Personal Assistant   
27630                           Senior Software Engineer   
27834  Executive Assistant (Office of President and C...   
14331                                          San Diego   
19323                                             Editor   
6969                                  Software Developer   
4264                                  Associate Attorney   
10948                           Senior Software Engineer   
8650                                           Developer   

                                       job_title_context salary(annualized)  \
26727                                       Receptionist              40000   
26531                                                NaN              47000   
766                                                  NaN             64,000   
752    I run a branch library and am responsible for ...             76,302   
11760                                                NaN            250,000   
4971                                                 NaN            534,300   
18930                                                NaN             38,760   
7739                 Patient care, teaching and research            145,000   
28054                                                NaN             300000   
9981                                                

There are 121 out of 205 unique values in the 'other_currency' column, with makes it a lot of unique values to deal with and the benefit of dealing with each one is minimal. So the current approach at the moment would be to first drop all the rows in the dataframe that have values in the other_currency column because they can skew the analysis of the data then drop the column itself after.

The other columns with high missing values (income_context, job_title_context) can be dropped without losing any relevant information.

In [10]:
# Remove rows where 'other_currency' column has non-null values.
df = df[df['other_currency'].isnull()]
df.shape

(27856, 18)

In [11]:
# Drop unnecessary columns.
df = df.drop(columns=['income_context', 'other_currency', 'job_title_context'],axis=1)

In [12]:
df.head()

timestamp age_range                       industry  \
0  4/27/2021 11:02:10     25-34   Education (Higher Education)   
1  4/27/2021 11:02:22     25-34              Computing or Tech   
2  4/27/2021 11:02:38     25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41     25-34                     Nonprofits   
4  4/27/2021 11:02:42     25-34  Accounting, Banking & Finance   

                                  job_title salary(annualized)  \
0        Research and Instruction Librarian             55,000   
1  Change & Internal Communications Manager             54,600   
2                      Marketing Specialist             34,000   
3                           Program Manager             62,000   
4                        Accounting Manager             60,000   

   additional_compensation currency         country        us_state  \
0                      0.0      USD   United States   Massachusetts   
1                   4000.0      GBP  United Kingdom             NaN   
2                      NaN      USD              US       Tennessee   
3                   3000.0      USD             USA       Wisconsin   
4                   7000.0      USD              US  South Carolina   

       us_city years_of_experience years_of_experience_in_field  \
0       Boston           5-7 years                    5-7 years   
1    Cambridge        8 - 10 years                    5-7 years   
2  Chattanooga         2 - 4 years                  2 - 4 years   
3    Milwaukee        8 - 10 years                    5-7 years   
4   Greenville        8 - 10 years                    5-7 years   

  level_of_education      gender   race  
0    Master's degree       Woman  White  
1     College degree  Non-binary  White  
2     College degree       Woman  White  
3     College degree       Woman  White  
4     College degree       Woman  White

**Cleaning other columns with missing values**

In [13]:
(df.isnull().mean() * 100).sort_values(ascending=False)

additional_compensation         25.969271
us_state                        17.428920
level_of_education               0.771827
race                             0.621051
gender                           0.610281
us_city                          0.294371
industry                         0.258472
job_title                        0.003590
salary(annualized)               0.000000
timestamp                        0.000000
age_range                        0.000000
currency                         0.000000
country                          0.000000
years_of_experience              0.000000
years_of_experience_in_field     0.000000
dtype: float64

- Additional Compensation Column

In [14]:
# fill missing values in 'additional_compensation' column with 0
df['additional_compensation'] = df['additional_compensation'].fillna(0)

- State(US) Column

In [15]:
# Viewing the pattern of the state(US) column to see if United States has have corresponding states and 
# non United States entries have null values.
df.iloc[:,6:9]

currency         country        us_state
0          USD   United States   Massachusetts
1          GBP  United Kingdom             NaN
2          USD              US       Tennessee
3          USD             USA       Wisconsin
4          USD              US  South Carolina
...        ...             ...             ...
28056      CAD          Canada             NaN
28057      USD  United States         Missouri
28058      USD             USA         Georgia
28059      USD         Myanmar        Colorado
28060      USD           Burma             NaN

[27856 rows x 3 columns]

In [16]:
# View rows with missing values in 'State(US)' column to confirm the pattern.
df[df['us_state'].isnull()].iloc[:,6:10]

currency          country us_state     us_city
1          GBP   United Kingdom      NaN   Cambridge
10         USD    United States      NaN  Boston, MA
14         CAD           Canada      NaN      Remote
15         GBP  United Kingdom       NaN     Lincoln
21         USD    United States      NaN     Atlanta
...        ...              ...      ...         ...
28050      EUR            Spain      NaN   Barcelona
28053      CAD           Taiwan      NaN      Taipei
28055      CAD           Canada      NaN     Toronto
28056      CAD           Canada      NaN   Kitchener
28060      USD            Burma      NaN      Yangon

[4855 rows x 4 columns]

It appears that there are some rows where the 'country' column is 'United States' but the 'State(US)' column is missing.

Since there are multiple entries where the in the country column United States is written in many different ways, we will convert all entries to 'United States' to make it consistent first before dealing with the missing state entries.

In [17]:
df['country'].unique()

array(['United States', 'United Kingdom', 'US', 'USA', 'Canada',
       'United Kingdom ', 'usa', 'UK', 'Scotland ', 'U.S.',
       'United States ', 'The Netherlands', 'Australia ', 'Spain', 'us',
       'Usa', 'England', 'finland', 'United States of America', 'France',
       'United states', 'Scotland', 'USA ', 'United states ', 'Germany',
       'UK ', 'united states', 'Ireland', 'Australia', 'Uk',
       'United States of America ', 'U.S. ', 'canada', 'Canada ', 'U.S>',
       'ISA', 'Great Britain ', 'US ', 'United State', 'U.S.A', 'Denmark',
       'U.S.A.', 'America', 'Netherlands', 'netherlands', 'England ',
       'united states of america', 'Ireland ', 'Switzerland',
       'Netherlands ', 'Bermuda', 'Us', 'The United States',
       'United State of America', 'Germany ', 'Mexico ', 'United Stated',
       'South Africa ', 'Belgium', 'Northern Ireland', 'u.s.',
       'South Africa', 'UNITED STATES', 'united States', 'Sweden',
       'Hong Kong', 'Sri lanka', 'Contracts', 'U

Starting with USA, we can see it was entered in many ways. We will convert them all to 'United States'.

- Convert all USA entries in 'country' column to 'United States'.

In [18]:
# First make all the entries in 'country' column lowercase.
df['country'] = df['country'].str.lower().str.strip()

In [19]:
# show all unique country entries that start with the letter 'u'
df[df['country'].str.startswith('u')]['country'].unique()

array(['united states', 'united kingdom', 'us', 'usa', 'uk', 'u.s.',
       'united states of america', 'u.s>', 'united state', 'u.s.a',
       'u.s.a.', 'united state of america', 'united stated',
       'usa-- virgin islands', 'united statws', 'u.s', 'unites states',
       'u. s.', 'united sates', 'united states of american',
       'uniited states', 'united kingdom (england)',
       'united sates of america',
       'united states (i work from home and my clients are all over the us/canada/pr',
       'unted states', 'united statesp', 'united stattes',
       'united statea', 'united kingdom.', 'united statees',
       'uniyed states', 'uniyes states', 'united states of americas',
       'u.a.', 'us of a', 'united arab emirates', 'u.k.', 'u.sa',
       'united kindom', 'united status', 'uxz', 'uss', 'uniteed states',
       'united stares', 'uk (northern ireland)', 'uk for u.s. company',
       'unite states', 'united kingdomk', 'unitedstates',
       'u.k. (northern england)', 'u

In [20]:
# Get all the usa varitions into list
usa_list = ['united states', 'us', 'usa', 'u.s.',
       'united states of america', 'u.s>', 'united state', 'u.s.a',
       'u.s.a.', 'united state of america', 'united stated',
       'usa-- virgin islands', 'united statws', 'u.s', 'unites states',
       'u. s.', 'united sates', 'united states of american',
       'uniited states', 'united sates of america',
       'united states (i work from home and my clients are all over the us/canada/pr',
       'unted states', 'united statesp', 'united stattes',
       'united statea', 'united statees',
       'uniyed states', 'uniyes states', 'united states of americas',
       'u.a.', 'us of a', 'u.sa',
       'united status', 'uniteed states',
       'united stares', 'unite states', 'unitedstates',
       'united statew', 'united statues', 'untied states',
       'usa (company is based in a us territory, i work remote)',
       'usab', 'unitied states',
       'united sttes', 'uniter statez', 'u. s',
       'usa tomorrow', 'united stateds',
       'usat', 'unitef stated', 'ua', 'usaa',
       'united y', 'united statss', 'united  states',
       'united states is america','america','isa','the united states',
       'san francisco', '🇺🇸','california','the us']

# replace all USA entries in 'country' column with 'United States' using the list usa_list
df['country'] = df['country'].apply(lambda x: 'united states' if x in usa_list else x)

In [21]:
df['country'].unique()

array(['united states', 'united kingdom', 'canada', 'uk', 'scotland',
       'the netherlands', 'australia', 'spain', 'england', 'finland',
       'france', 'germany', 'ireland', 'great britain', 'denmark',
       'netherlands', 'switzerland', 'bermuda', 'mexico', 'south africa',
       'belgium', 'northern ireland', 'sweden', 'hong kong', 'sri lanka',
       'contracts', 'england/uk',
       "we don't get raises, we get quarterly bonuses, but they periodically asses income in the area you work, so i got a raise because a 3rd party assessment showed i was paid too little for the area we were located",
       'england, uk.', 'greece', 'japan', 'britain', 'austria',
       'canada, ottawa, ontario', 'global', 'united kingdom (england)',
       'worldwide (based in us but short term trips aroudn the world)',
       'canadw', 'luxembourg', 'united kingdom.', 'new zealand',
       'cayman islands', 'can',
       'i am located in canada but i work for a company in the us',
       'latvia', '

In [22]:
# Get all the uk varitions into list
uk_list = ['uk', 'united kingdom (england)', 'united kingdom'
       'united kingdom.', 'u.k.', 'united kindom',
       'uk (northern ireland)', 'uk for u.s. company',
       'united kingdomk', 'u.k. (northern england)', 'u.k',
       'uk (england)', 'uk, remote', 'unites kingdom',
       'uk, but for globally fully remote company','england','england/uk',
       'england, uk.','britain','england, uk','wales (united kingdom)', 'england, gb',
       'england, united kingdom','englang','scotland, uk','wales, uk', 'northern ireland',
       'wales (uk)','northern ireland, united kingdom','london','great britain']

# replace all USA entries in 'country' column with 'United States' using the list usa_list
df['country'] = df['country'].apply(lambda x: 'united kingdom' if x in uk_list else x)

In [23]:
# Get all the canada varitions into list
canada_list = ['canada', 'canada, ottawa, ontario', 'canadw',
       'can', 'canda', 'canada and usa', 'canad',
       'canadá','csnada']

# replace all USA entries in 'country' column with 'United States' using the list usa_list
df['country'] = df['country'].apply(lambda x: 'canada' if x in canada_list else x)

Since some entries in the 'country' column have sentences rather than country names, we will drop these rows

In [24]:
# Drop rows where 'country' column has more than 3 words
df = df[df['country'].str.split().apply(len) <= 3]

Getting the count of each country in the dataframe

In [25]:
value_counts = df['country'].value_counts()
value_counts

country
united states     23141
canada             1683
united kingdom     1544
australia           377
germany             195
                  ...  
saudi arabia          1
loutreland            1
taiwan                1
myanmar               1
burma                 1
Name: count, Length: 126, dtype: int64

Also based on the frequency of the top countries, we will keep only the countries with more than 100 entries, as countries with less than that may not provide meaningful insights in this dataset.

In [26]:
df = df[df['country'].isin(value_counts[value_counts > 100].index)]

**Now that we are done cleaning the 'country' column, we can now go back to the state column to fill the missing values.**

Fisrt we will check the rows where the 'country' column is 'United States' but the state column is missing.

In [27]:
df.query('country == "united states" and us_state.isnull()')

timestamp age_range                              industry  \
10     4/27/2021 11:03:03     25-34                            Nonprofits   
21     4/27/2021 11:03:24     35-44                            Nonprofits   
71     4/27/2021 11:04:39     35-44                 Aerospace contracting   
132    4/27/2021 11:05:46     35-44                     Computing or Tech   
133    4/27/2021 11:05:46     55-64          Education (Higher Education)   
...                   ...       ...                                   ...   
25323   5/6/2021 12:39:18     35-44  Government and Public Administration   
25599    5/7/2021 1:18:37     25-34                           Health care   
26119  5/12/2021 23:53:52     35-44                       Media & Digital   
26494  6/22/2021 11:47:22     35-44         Education (Primary/Secondary)   
28006    2/6/2024 9:07:45  under 18                                   NaN   

                                     job_title salary(annualized)  \
10                              Office Manager             47,500   
21     Program Coordinator & Assistant Editor              50,000   
71                           Project engineer             122,000   
132                            Product Manager            130,000   
133               career services professional             55,000   
...                                        ...                ...   
25323                    Senior Fiscal Analyst             130000   
25599                      Physician Assistant             135000   
26119                                   Editor              45000   
26494                      High School Teacher              56650   
28006                                       vv                 76   

       additional_compensation currency        country us_state  \
10                         0.0      USD  united states      NaN   
21                         0.0      USD  united states      NaN   
71                      3600.0      USD  united states      NaN   
132                        0.0      USD  united states      NaN   
133                        0.0      USD  united states      NaN   
...                        ...      ...            ...      ...   
25323                  25000.0      USD  united states      NaN   
25599                      0.0      USD  united states      NaN   
26119                      0.0      USD  united states      NaN   
26494                   1000.0      USD  united states      NaN   
28006                      0.0      USD  united states      NaN   

                                    us_city years_of_experience  \
10                               Boston, MA           5-7 years   
21                                  Atlanta           5-7 years   
71                         Colorado Springs       11 - 20 years   
132                           Philadelphia        11 - 20 years   
133                            Indianapolis       21 - 30 years   
...                                     ...                 ...   
25323                Prefer not to disclose       11 - 20 years   
25599                           Springfield         2 - 4 years   
26119  Fully remote company based out of CA       11 - 20 years   
26494                          Philadelphia           5-7 years   
28006                                irving      1 year or less   

      years_of_experience_in_field level_of_education gender   race  
10                       5-7 years     College degree  Woman  White  
21                     2 - 4 years                PhD  Woman  White  
71                   11 - 20 years    Master's degree  Woman  White  
132                  11 - 20 years     College degree    Man  White  
133                  21 - 30 years    Master's degree  Woman  White  
...                            ...                ...    ...    ...  
25323                11 - 20 years    Master's degree  Woman  White  
25599                  2 - 4 years    Master's degree  Woman  White  
26119                 

- We will first deal with the entries where the 'country' column is 'United States' and the state column is missing and the currency column is 'usd' which verifies that this are actual 'united states' entries. We will fill this with "ND" for "Not Disclosed".

- Finally, we will fill correctly identified non united states entries with "N/A" for "Not Applicable".

In [28]:
# Fill entries where 'country' column is 'United States' and 'us_state' column is missing with 'ND'
df.loc[
    (df['country'] == "united states") & (df['us_state'].isnull()) & (df['currency'] == "USD"),
    'us_state'
] = "ND"


In [31]:
# Making sure the remaining missing values are not in 'United States'
df[df['us_state'].isnull()].sample(5)

timestamp age_range                              industry  \
11898   4/28/2021 6:40:30     35-44           Marketing, Advertising & PR   
11713   4/28/2021 4:00:59     35-44                     Computing or Tech   
1734   4/27/2021 11:31:03     35-44  Government and Public Administration   
26461  6/13/2021 13:09:36     35-44                                Retail   
22903    5/1/2021 7:13:41     25-34                Transport or Logistics   

                   job_title salary(annualized)  additional_compensation  \
11898      Head of Marketing             35,000                      0.0   
11713      Junior Developer              26,500                      0.0   
1734   Senior Policy Analyst            115,000                      0.0   
26461           Book shelver              35000                      0.0   
22903                Planner              27000                      0.0   

      currency         country us_state              us_city  \
11898      GBP  united kingdom      NaN  Newcastle-upon-Tyne   
11713      GBP  united kingdom      NaN              Bristol   
1734       CAD          canada      NaN               Ottawa   
26461      CAD          canada      NaN              Calgary   
22903      GBP  united kingdom      NaN          Manchester    

      years_of_experience years_of_experience_in_field level_of_education  \
11898       11 - 20 years                 8 - 10 years     College degree   
11713           5-7 years               1 year or less     College degree   
1734        11 - 20 years                11 - 20 years     College degree   
26461       11 - 20 years                11 - 20 years     College degree   
22903       11 - 20 years                  2 - 4 years       Some college   

                              gender  \
11898                            Man   
11713                          Woman   
1734                           Woman   
26461  Other or prefer not to answer   
22903                          Woman   

                                                    race  
11898                                              White  
11713                                              White  
1734   Another option not listed here or prefer not t...  
26461                                              White  
22903                                              White

In [32]:
# Fill remaining missing values in 'us_state' column with 'N/A'
df['us_state'] = df['us_state'].fillna("N/A")